[P092] Reduce memory usage when plugin not used

uwekaditz opened the pull request:
Reduce memory usage when plugin not used (see PR #3248)
CHG: all functions and variables moved to P092_data_struct()
CHG: keeping only global values needed for all tasks
CHG: description in P092_DLbus.rst that each value read from the DL bus needs a new device task

**Note** that each reading from the DL bus requires a **new** device task.
But the data structure should be created and initialized only once and each task has to point to the same structure.
The reading and decoding of the DL bus happens also only once.

uwekaditz committed:
[P092] Reduce memory usage when plugin not used PR #3248

CHG: all functions and variables moved to P092_data_struct()
CHG: keeping only global values needed for all tasks
CHG: description in P092_DLbus.rst that each velue need a new device task

TD-er commented:
So this PR supersedes my PR?
Meaning I can close #3248 ?

uwekaditz commented:
Yes, you can close your PR.

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
BUG: #endif missing in P092_data_struct.h

TD-er commented on src/src/PluginStructs/P092_data_struct.h:
```diff
@@ -1,5 +1,5 @@
-#ifndef DLBus_H
-#define DLBus_H
+#ifndef PLUGINSTRUCTS_P036_DATA_STRUCT_H
+#define PLUGINSTRUCTS_P036_DATA_STRUCT_H
```
Please use P092 instead of P036, or else you will get very strange compile results when using both plugins :)

uwekaditz committed:
Copy and Paste error

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz commented:
@TD-er Should I integrate the changes of PR #3345 into my PR or will it be done by @pez3? The changes are important because I have only an ESR21 for testing. All the other devices are just coded based on their documentations.

TD-er commented:
Yes please, so we can make a test build first.
This can then be tested.

uwekaditz committed:
integrate the changes of PR #3345 (created by pez3)

CHG: removed internal pullup (at least for devices UVR61-3 (V8.3) upwards)
BUG: fixed setting of interrupt. so changing of GPIO should work now
BUG: fixed decoding of UVR61-3 frames. UVR61-3 (V8.3) devices have two analog outputs

TD-er commented on src/src/PluginStructs/P092_data_struct.cpp:
```diff
@@ -462,7 +462,7 @@ P092_data_struct::~P092_data_struct() {
 }
 
-bool P092_data_struct::init(int8_t pin1) {
+bool P092_data_struct::init(int8_t pin1, int DeviceIndex) {
```
Is it possible to use a different name for this variable?
In ESPEasy we also have a DeviceIndex, which means something completely different.

uwekaditz replied:
I have changed the variable name to P092DeviceIndex (see last commit)

pez3 commented on src/_P092_DLbus.ino:
```diff
@@ -337,9 +267,10 @@ boolean Plugin_092(uint8_t function, struct EventStruct *event, String& string)
           // interrupt was already attached to P092_DLB_Pin
-          detachInterrupt(digitalPinToInterrupt(DLbus_Data->ISR_DLB_Pin));
+          detachInterrupt(digitalPinToInterrupt(P092_data->DLbus_Data->ISR_DLB_Pin));
+          P092_data->DLbus_Data->IsISRset = false;  //to ensure that a new interrupt is attached in P092_data->init()
```
I'm not sure whether this is the optimal place to set the IsISRset to false.

I'm not that familiar with timings of esp easy but since the new interrupt is attached in PLUGIN_ONCE_A_SECOND i think under bad circumstances PLUGIN_ONCE_A_SECOND could be called before PLUGIN_INIT (the place where the new pin is set) and then the interrupt is not attached to the correct pin.

Therefore i suggested to place setting IsISRset to false before setting the new pin during initialization (where you already added it ;) )

TD-er replied:
Yep, I think that's a valid point.
Reset the flag first before the detachInterrupt.
And maybe also place it in a separate scope so the compiler may not perform an out-of-order optimization.
```c++
{
  P092_data->DLbus_Data->IsISRset = false;  //to ensure that a new interrupt is attached in P092_data->init()
}
detachInterrupt(digitalPinToInterrupt(P092_data->DLbus_Data->ISR_DLB_Pin));
```

pez3 replied:
i would completely delete it here because its already done in init (triggered by flag P092_init).

TD-er replied:
Hmm for that I may have to look a bit more into it...

pez3 commented on src/_P092_DLbus.ino:
```diff
@@ -423,34 +354,38 @@ boolean Plugin_092(uint8_t function, struct EventStruct *event, String& string)
       }
       else {
-        addLog(LOG_LEVEL_INFO, F("P092_init ..."));
+        if (P092_data == nullptr) {
```
i'm not quite sure whether re-initialization is done in case GPIO pin is changed because of this query.

P092_init is set to true in PLUGIN_WEBFORM_SAVE but i think P092_data is still != nullptr and therefore the initialization function is not called.

pez3 commented on src/_P092_DLbus.ino:
```diff
@@ -366,7 +372,7 @@ boolean Plugin_092(uint8_t function, struct EventStruct *event, String& string)
             addLog(LOG_LEVEL_INFO, F("Init P092_data_struct ..."));
 
-            if (!P092_data->init(CONFIG_PIN1)) {
+            if (!P092_data->init(CONFIG_PIN1, PCONFIG(0))) {
```
I'm not quite sure whether re-initialization (and therefore new GPIO configuration) is done when GPIO is changed in webform. P092_init is set to false in PLUGIN_WEBFORM_SAVE but i think the initialization function is never called because P092_data is != nullptr in this case.

uwekaditz replied:
I have reviewed the use of P092_init (see last commit)

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
[P092] P092_data->init() is always done if P092_init == false

CHG: P092_data->init() is always done if P092_init == false, not depending on P092_data == nullptr
CHG: changes variable name DeviceIndex to P092DeviceIndex

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Update P092_data_struct.cpp

pinMode can not be used for ESP32 (no Input_PullDown option), must be set on the hardware page

uwekaditz commented:
@TD-er Will be there an option for the `Input_PullDown `for the ESP32-GPIO used in a plug in in near future?

uwekaditz committed:
[P092] PinMode in Web interface adjustable (even InputPullDown if available)

CHG: Pinmode in Web interface adjustable (even InputPullDown if available for ESP32)
CHG: old pinmode settings (UVR61-3 without pullup, all other types with pullup) are kept
CHG: code simplified (ValueCount has to be 1)

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
wrong comment sign

TD-er commented on src/src/PluginStructs/P092_data_struct.cpp:
```diff
@@ -434,7 +434,7 @@ boolean DLBus::CheckCRC(uint8_t IdxCRC) {
 }
 
-// sensor types
+/ sensor types
```
Huh?
I don't get this.
Change from a comment to something that will not compile.

uwekaditz replied:
My mistake, already changed in the commit c6a3468

uwekaditz committed:
[P092] Documentation for UVR61-3 (v8.3 or higher) added

tonhuisman commented on docs/source/Plugin/P092_UVR61-3v8_3.rst:
```diff
@@ -0,0 +1,85 @@
+* The data frame consist of 61 bytes (excluding sync and CRC).
+* The data are transmitted @ 488Hz, therefore the transmission of one data set takes about 1.3 seconds.
+
+.. image:: P092_UVR61-3.png
```
Are these images (more below) already committed earlier? As I can't find them all in this PR...

uwekaditz replied:
Yes, they are already in the latest release.

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz commented:
I'm done.
From my side everything is ready.
The changes TD-er requested have been done on Oct, 28th.
The documentation for UVR61-3 (v8.3 or higher) has been added.

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz
